psi4/driver/procrouting/scf_proc/diis.py

```diff
@@ -0,0 +1,200 @@
+
+    def build_entry(self, entry, target_index):
+        assert len(self.R_template) + len(self.T_template) == len(entry)
+        R = entry[:len(self.R_template)]
+        T = entry[len(self.R_template):]
+
+        return [
+                [self.copier(Ri, self.get_name("R", target_index, i)) for i, Ri in enumerate(R)],
+                [self.copier(Ti, self.get_name("T", target_index, i)) for i, Ti in enumerate(T)]
+        ]
+
+    def add_entry(self, *args):
+        if self.max_vecs == 0:
+            return False
+
+        self.last_added += 1
+        self.last_added % self.max_vecs
+
+        if len(self.stored_vectors) >= self.max_vecs:
+            if self.removal_policy == RemovalPolicy.OldestAdded:
+                target_index = self.last_added
+            elif self.removal_policy == RemovalPolicy.LargestError:
+                target_index = np.argmax([self.get_dot_product(i, i) for i in range(len(self.stored_vectors))])
+            else:
+                raise Exception(f"RemovalPolicy {self.removal_policy} not recognized. This is a bug: contact developers.")
+            # Purge imminently-outdated values from cache.
+            self.cached_dot_products = {key: val for key, val in self.cached_dot_products.items() if target_index not in key}
+            # Set the new entry.
+            self.stored_vectors[target_index] = self.build_entry(args, target_index)
+        else:
+            self.stored_vectors.append(self.build_entry(args, self.last_added))
+
+        return True
+
+    def extrapolate(self, *args):
+
+        dim = len(self.stored_vectors) + 1
+        B = np.zeros((dim, dim))
+        for i in range(len(self.stored_vectors)):
+            for j in range(len(self.stored_vectors)):
+                B[i, j] = self.get_dot_product(i, j)
+        B[-1, :-1] = B[:-1, -1] = -1
+        # Normalize for numerical stability
+        max_val = np.abs(B[:-1, :-1]).max()
+        if max_val:
+            B[:-1, :-1] /= max_val
+
+        rhs = np.zeros((dim))
+        rhs[-1] = -1
+        coeffs = np.linalg.lstsq(B, rhs, rcond=None)[0][:-1]
+
+        for j, Tj in enumerate(args):
+            Tj.zero()
+            if self.storage_policy == StoragePolicy.InCore:
+                for ci, (_, Ti) in zip(coeffs, self.stored_vectors):
+                    axpy(Tj, ci, Ti[j])
+            elif self.storage_policy == StoragePolicy.OnDisk:
+                for i, ci in enumerate(coeffs):
+                    psio = core.IO.shared_object()
+                    if isinstance(Tj, core.Vector):
+                        Tij = core.Vector(self.get_name("T", i, j), *self.T_template[j])
+                        Tij.load(psio, psif.PSIF_LIBDIIS)
+                    elif isinstance(Tj, (core.Matrix, core.dpdfile2, core.dpdbuf4)):
+                        Tij = core.Matrix(self.get_name("T", i, j), *self.T_template[j])
+                        Tij.load(psio, psif.PSIF_LIBDIIS, core.SaveType.SubBlocks)
```

Comment:
Does this load all the data into memory? Is that safe to assume in all cases?

Reply:
It loads all the data in memory, yes. The existing code makes even more optimistic memory assumptions: that you can load [all items at once](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libdiis/diismanager.cc#L365-L366). Removing this assumption is a feature addition, rather than a migration.